### PR TITLE
fix(router): handle JSON connector response parse error

### PR DESCRIPTION
**Pull request by AkshayaFoiger**
## Type of Change

- [x] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Certain connectors (Braintree, Rapyd, Trustpay, Boku) send error response in HTML or string format. This lead to a ``HE_00`` in our system as we try to accept it in a JSON format. Now we have introduced a ``get_json_deserialized`` function in router/utils.rs that can  be used to handle Json error. Json Error handling is implemented for Trustpay, Rapyd and Braintree. Boku already handles it. 

## How did you test it?

Screen Shot 2023-08-08 at 5 16 42 PM

Screen Shot 2023-08-08 at 8 16 57 PM

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/trustpay.rs [hunk 1]
```diff
@@ -104,34 +104,43 @@ impl ConnectorCommon for Trustpay {
...
+                        .unwrap_or(consts::NO_ERROR_CODE.to_string()),
+                    // message vary for the same code, so relying on code alone as it is unique
+                    message: option_error_code_message
+                        .map(|error_code_message| error_code_message.error_code)
+                        .unwrap_or(consts::NO_ERROR_MESSAGE.to_string()),
+                    reason: reason.or(response_data.description),
+                })
+            }
+            Err(error_msg) => {
+                crate::logger::error!("{}", error_msg);
```
Comment (ArjunKarthik) [inline crates/router/src/connector/trustpay.rs:140]:
```suggestion
                crate::logger::error!(deserialization_error= error_msg);
```

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/braintree.rs [hunk 2]
```diff
@@ -46,6 +46,27 @@ impl ConnectorCommon for Braintree {
...
+
+        match response {
+            Ok(response_data) => Ok(types::ErrorResponse {
+                status_code: res.status_code,
+                code: consts::NO_ERROR_CODE.to_string(),
+                message: response_data.api_error_response.message,
+                reason: None,
+            }),
+            Err(error_msg) => {
+                crate::logger::error!("{}", error_msg);
```
Comment (jarnura) [inline crates/router/src/connector/braintree.rs:65]:
```suggestion
                logger::error!("{}", error_msg);
```

File: crates/router/src/connector/rapyd.rs [hunk 3]
```diff
@@ -82,16 +82,23 @@ impl ConnectorCommon for Rapyd {
...
+
+        match response {
+            Ok(response_data) => Ok(ErrorResponse {
+                status_code: res.status_code,
+                code: response_data.status.error_code,
+                message: response_data.status.status.unwrap_or_default(),
+                reason: response_data.status.message,
+            }),
+            Err(error_msg) => {
+                crate::logger::error!("{}", error_msg);
```
Comment (jarnura) [inline crates/router/src/connector/rapyd.rs:98]:
```suggestion
                logger::error!("{}", error_msg);
```

File: crates/router/src/connector/trustpay.rs (hunk 1 above)
Comment (jarnura) [inline crates/router/src/connector/trustpay.rs:140]:
```suggestion
               logger::error!("{}", error_msg);
```

File: crates/router/src/utils.rs [hunk 4]
```diff
@@ -176,3 +179,35 @@ mod tests {
...
+    let response_data = String::from_utf8(res.response.to_vec())
+        .into_report()
+        .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+
+    // check for whether the response is in json format
+    match serde_json::from_str::<Value>(&response_data) {
+        // in case of unexpected response but in json format
+        Ok(_) => Err(errors::ConnectorError::ResponseDeserializationFailed)?,
+        // in case of unexpected response but in html or string format
+        Err(_) => {
```
Comment (jarnura) [inline crates/router/src/utils.rs:203]: Log the Error also

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by jarnura — APPROVED** [review]

**Pull request merged**